src/Npgsql/NpgsqlEventSource.cs

```diff
@@ -174,7 +181,7 @@ protected override void OnEventCommand(EventCommandEventArgs command)
                     DisplayUnits = "%"
                 };
 
-                _poolsCounter = new PollingCounter("connection-pools", this, () => _pools)
+                _poolsCounter = new PollingCounter("connection-pools", this, () => _pools.Count)
```

Comment:
Doesn't this need to lock too?

Reply:
In the current implementation of HashSet, the Count property simply returns the value of the internal variable m_count.
But maybe this could be changed in future versions. So I added lock just now.

Reply:
Or we can just add a count variable to avoid locking completely. But do not think it's that important.

Reply:
Guys, it's just an atomic memory read even on 32-bit machine.

Reply:
@YohDeadfall as @neyromant wrote, HashSet doesn't guarantee that this property is thread-safe - so why not be safe and lock? Are you (or @vonzshik) seeing any perf issue here that justifies taking risks, given that this lambda can get called only once per second at most?

Reply:
Atomic reads are always thread safe, but not other operations on the collections. We all know how it works, it's open source, and we also all know that it's the best possible way to implement the property. No one will touch it.

My opinion here that `lock`, while theoretically should be here, should not be used. It gives nothing here, really.

Reply:
@YohDeadfall nothing in the [HashSet documentation for Count](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.count?view=net-5.0#System_Collections_Generic_HashSet_1_Count) says that the property is implemented as an atomic read. That means the implementation could change in the future, and our code would start breaking.

> My opinion here that lock, while theoretically should be here, should not be used. It gives nothing here, really.

I think the question is the opposite one - what would be our reason for *not* using a lock, and thereby ensuring everything always works in the future even if the HashSet.Count implementation changes? What is the downside, given that this property is called once per second?

Reply:
Knowing the area and Microsoft well, I surely can say changes are very unexpected here. It might happen only if someone makes a faster less memory consuming hash set, but it's how the hash set works now in all frameworks.

Feel free to ignore my opinion here, but I still think that it's useless here.

Reply:
> @YohDeadfall as @neyromant wrote, HashSet doesn't guarantee that this property is thread-safe - so why not be safe and lock? Are you (or @vonzshik) seeing any perf issue here that justifies taking risks, given that this lambda can get called only once per second at most?

I do not see any perf issues here. But if we can make it faster, while also not depending on `HashSet.Count` implementation, I do not see a reason why not do so, especially when it doesn't cost us anything.

Reply:
I agree with you that there's little chance this will change - but it's always better not to code assumptions like this which could break us later (as a matter of principle). Dictionary underwent some very serious changes in .NET Core, it could happen with HashSet too.

I don't want to ignore your opinion... But the more important answer I'm looking for, is what is the *downside* of using the lock (even if it is useless), given that this code is very cold. If you're seeing some scenario where this could impact us badly, I'd want to know about it... If there isn't one, then why not be safer?

Reply:
The only difference between `Dictionary<,>` and `HashSet<>` is ability to store values. Otherwise, they are identical, both a hash sets with mostly the same set of methods. The chance is near to be zero since properties being methods under the hood should conform to execution cost requirement of `O(1)` and not perform any computations in case of getting a value.

There's a chance to face a multitenant application dealing with time limited passwords. In that case it might be critical, but otherwise, no downsides except that the lock takes much more time than the actual read.

Reply:
> There's a chance to face a multitenant application dealing with time limited passwords. In that case it might be critical, but otherwise, no downsides except that the lock takes much more time than the actual read.

Regardless of multitenant applications or time-limited passwords, the counter is polled once per second - that's as cold as can be. Any added overhead for the lock is going to be negligible.